### ISSUE TITLE
Advanced task dependencies / caching

### ISSUE THREAD

**Issue by bitprophet** [pyinvoke/invoke#461]
## Use case description

Core desire is to strengthen relationships between tasks above/beyond what the existing pre/post functionality achieves; and to structure a nontrivial set of dependent state as a graph of tasks which only execute as many times as is required to set up that state.

Specific example is an existing nontrivial task tree around performing cloud automation, configuration management, and similar things. Many/most tasks in this tree are currently wrapped with a `@state` decorator whose body does a large amount of "heavy lifting":

- processes pre-existing configuration directives, including aborting if they aren't defined
  - A reminiscent predecessor is the old and (as far as I can tell) barely used [fabric.api.require](http://docs.fabfile.org/en/1.13/api/core/operations.html#fabric.operations.require)
- instantiate network and API clients using that early configuration (and/or pulling from a secrets storage API)
- generate new configuration settings based on runtime information (including data gathered from the above clients)
- attempts at each step to perform idempotency, including skipping expensive operations if necessary (so no re-connecting to already-connected APIs; no re-reading already-read files; etc.)

At the least, this decorator "wants" to be split up into a bunch of smaller parts, each of which are connected by what state they themselves depend on, with the outermost/topmost caller (i.e. the decorated CLI task) specifying more precisely which bits they need. (This already exists in part with some args passed into the decorator, e.g. `@state(limited=True)` skips everything but the first few config bits and an early API client.)

Those smaller parts would then be capable of memoizing their results, so that e.g. if they are declared to "generate a value for config path `c.aws.clients.vpc`", and that config path already exists/is non-empty when something starts calling that function, it simply records this fact and short-circuits.

Presumably one could expand this concept all the way to abstract things like "task A requires an admin cloud instance and a database" - including some sort of registry-driven "find me some task that satisfies needs XYZ, cuz I need X Y and Z" setup (though that feels much more fraught and complex, and is probably a distinct API feel from the "do exactly what i say but feel free to skip anything you already did" described above.)

## Solutions brainstorm

TK

----

Related, possibly subsumed tickets:

- #41 - the original dependencies ticket, long closed
- #45 - this is the closest match, and arguably a duplicate, though I interpret that ticket to be more "make the existing pre/post deduplicating use a DAG", whereas this ticket is about extending pre/post itself to have a richer API, even if that probably requires the DAG anyways.
- #100 - if we grow the ability to say "this task should generate <file X>", it makes a ton of sense to slightly extend that to "this task should generate <file X>, regenerating if <file X> exists but is old"
- #170 - insofar as calling tasks from other tasks definitely wants to honor this new functionality

**Comment by bitprophet** [comment]
More thoughts...

## Call caching/tracking

Any non-empirical task call tracking should be double-checked for eventual inter-process parallelization safety as well as intra-process parameterization safety:

- Inter-process isn't as big a deal yet since current parameterization (eg Fabric 2 hosts stuff) focuses on threading, since we're thread-safe; but we really don't want to paint ourselves into a corner.
- Intra-process matters more immediately - e.g. if one has a host-oriented Fabric task which declares some dependency, and its dependency deduplication requires call tracking instead of state checking, we wouldn't want the state stored on the `Call` objects, but on the source `Task` object (or the `Collection` namespace obj). Otherwise each per-host `Call` would attempt to rerun its dependencies.
  - Conversely, though, we don't want to limit use cases where a parameterized `Task` _does_ want each generated `Call` to run its dependency (e.g. if the parameterization matters for the dependent task(s) as well as the main one - especially true in the "dependencies get a copy of the main task's args and/or context" case from #261 

## Decorator API

I'm now rethinking the "separate decorator" API option, for a couple reasons:

- First, adding decorators besides `@task` creates a decorator-ordering problem - see e.g. how Fabric 1 struggled/struggles with that to this day, with folks putting `@task` above and/or below things like `@roles`, `@hosts`, `@runs_once` etc.
  - Nontrivial mental overhead to remember which order is "allowed", especially for newbies, but even for intermediate users it's a footgun.
  - Requires extra code if one wants to try avoiding some of the ordering issue (as Fab 1 did) - all decorators must be "aware" of each other, add attributes to the decorated object and/or look for those added attributes, make sure not to nuke them when transforming from func to `Task`, etc.
  - Even with that extra code, it's not perfect, and now there's more opportunity for bugs. Such bugs are also harder to troubleshoot due to their nature (decorator-hosted info silently disappears, tasks are no longer created, etc.)
- Extra kwargs in `@task` aren't, in my current experience, a huge deal either - many use cases only make use of 1-2 kwargs per task (often 0) so they don't really end up being big ugly globs of kwargs.
  - And even if they did, it really is just an aesthetics thing - you're still talking maximum of 1 line per 'dimension', and for the more trivial args, they can often remain an easily readable 1-liner even w/ multiple in play (e.g. `@task(default=True, dependencies=(foo, bar))`)

**Comment by bitprophet** [comment]
Re: how to track the state: probably a good time to look at an actual graph lib. The one that Bruce identified and updated to be Python 3 compatible over in #45 is a good place to start. Poking it briefly (including a skim of [its docs](https://github.com/BruceEckel/pythondagger23/blob/master/doc/index.rst)):

- It's explicitly file-oriented by default, which is good for folks who want that, but implies work needed for other use cases (which, for my current real world test case, is all of them)
- Not very Pythonic code, oh well.
  - Lots of lowercase-named classes, defining its own iterator class that doesn't implement the iterator protocol (and which isn't returned by `__iter__` on the main dag object) etc.
- Effectively a single-file module, in terms of what is required for vendoring.
  - Most files present in the repo are administrivia (benchmarks, docs, tests) or appear to be junk ('old' copies of files, a rendered copy of the sphinx docs, etc)
- Primary API/flow:
  - Create a `dagger()` object, with optional persistence options (eg store hashes in a text file, or in a sqlite db [optionally in-memory])
  - Call `dagger.add('filepath', ['dependencies', 'here'])` to add nodes and their deps
  - Tell it to scan the files for staleness and then evaluate the necessary required calls to get everything un-stale, via `dagger.run()`
  - Obtain an iterator (`dagger.iter()`) and then iterate it with repeated calls to `iterator.next()` to walk the DAG (`next()` yields node objects with path info etc), ending when all nodes are non-stale.
- Secondary API:
  - Introspection:
    - `dagger.dump()`: Textual linebreak-separated breakdown of all nodes & their current staleness/hash/etc (though not their dependency relationships)
    - `dagger.tree()`: Another textual display, this one with just the relationships.
    - `dagger.dot()`: Graphviz `.dot` syntax, w/ optional write-to-file for display via any compatible app (e.g. graphviz itself, `dot -Tpng some.dot > some.png`)
    - `dagger.ordernames()`: textual comma-separated list (i.e. `sub-dep,other-sub-dep,dep,other-dep,top`)
    - `dagger.order` yields an `idict` object which is the underpinning of `ordernames()` and `iter()`, it exposes a regular old list accessed by `.list`, or a `name->index` mapping accessed by `.dict`
  - Node manipulation/metadata:
    - Can explicitly mark nodes (and thus, all of their dependents) as being out of date via `dagger.stale(name)`. (This seems to need to be done before `run()`; i.e. the workflow is probably 'static' instead of derived, but given the emphasis on speed, this makes sense.)
    - Can mark a node as "phony" (via `dagger.phony(name)`) which seems to just make it not bother performing file existence/date/hash check, or rather, allows one to override that default check.
      - Not 100% clear why that's useful (unless it's sort of an inverse of `.stale()`) but presumably something we'd have to tussle with to get non-file-related junk working, period
    - `dagger.exporthash` writes out the hash database to disk (depending on the format specified at instantiation)
    - Has the (IMHO) antipattern of "`get(name)` implicitly creates a new node and returns it"
    -

**Comment by bitprophet** [comment]
So, after a source code skim, dagger is one of those classic "is it even worth saving myself <=375 SLOC if I am gonna feel compelled to tweak and/or subclass a bunch of stuff" gray area cases. Might as well use it to prototype for now but I reserve the right to just write an "inspired by" recreation if I start running into too much trouble.

**Comment by bitprophet** [comment]
Been wondering during all this if we can reasonably do away with post-tasks. They're a very minor use case compared to their inverse, have difficult naming, and don't fit neatly into the idea of a DAG: they're not part of the DAG at all but are part of the "body" of the top-level/being-invoked task, practically speaking. (And you can't phrase them as inverse dependencies, either, since the whole point is that they are _not_ the focus of the execution.)

Further, most use cases for post-tasks seem like they hinge on success vs failure, which combined with the above, feels like they "should" be treated as part of #170 (tasks calling other tasks) and wrapped within `try`/`except`/`finally` blocks.

----

I scanned the tracker to see how many folks are filing tickets about post (vs pre-only) tasks and did find #298, so at least some users are using them - tho that is the only ticket I found.

It raises some things not noted prior:

The old deduplication was generally confusing re: how it treated post-tasks. tl;dr when post-tasks would show up multiple times, only the first one was kept, instead of only the last.

Corollary: even in a DAG, deduplication may want to be its own thing, because "lazy" or "elastic" pre/post tasks (what 'dependencies' generally mean - "run me _at least once_ per session, _anytime_ before/after the declaring task) are distinct from "rigid" ones (i.e. "run me _once per declaring task_" and typically also "run me _immediately_ before/after the declaring task").

That said, I think it can still be argued that "rigid" pre/post tasks can/should be phrased within the body of the main task and don't benefit from the outer dependency/dedupe system (in fact they don't benefit from _and significantly complicate_ such a system!)

**Comment by bitprophet** [comment]
If we assume we're gonna handle post-tasks, kind of want something that more closely matches the term `dependency`/`depends`/`requires`. Brainstorm:

- `triggers`, i.e. execution of the main task triggers (eventual) execution of another afterwards
- `drives`. Meh.
- `after` is tempting, but also far too ambiguous due to lack of clear object/subject - does `@task(after=[foobar])` mean "run `foobar` after you run this task" _or_ "run this task after `foobar` runs"? It's not clear without prior knowledge that it's being used to implement this particular functionality.
  - Could potentially use `afterwards` (`@task(afterwards=foobar)`) which is less ambiguous ("run me, then afterwards, run `foobar`) but still doesn't feel right to me, too awkward and/or cutesy?
- `later`, as in, "run this task, then later, run `foobar`. Also awkward/cutesy.
- Not sure what else works?

**Comment by indera** [comment]
Travis uses `after_success` or `after_failure` ... https://docs.travis-ci.com/user/customizing-the-build

**Comment by bitprophet** [comment]
@indera Indeed, and the thought occurred to me, though I suspect in most cases folks will desire to handle the `after_failure` case 'internally' with `try`/`except`/`finally` type logic as opposed to reimplementing that functionality within the execution system. We'll see - it's certainly possible to add the success/failure split later.

**Comment by bitprophet** [comment]
A TODO for myself to get this off the ground and stop dithering:

- [ ] Finish the DDD I'm writing in the conceptual docs. It's turning into a restatement of what was there before, but oh well. Still includes the new ideas...
- [ ] Implement the DAG side of things, without otherwise touching `@task` or changing too much behavior - i.e. take the step of replacing the existing dedupe with `dagger`.
  - In this case, I think we have to start out marking all nodes `stale` right after adding, and then `remove` them after they are executed.
- [ ] Next is probably a good time to see about fixing #298, presumably by adding post-tasks to the runlist being fed to the DAG, as late as possible (instead of as early as possible)
  - (A new day, a new brain) _Can_ we simply phrase post-tasks as tasks added to the DAG as depending-on the actual runtime runlist?
  - I don't really see another way of handling it besides running 2 or more DAG sessions (one for the main runlist, then another 1..N for any post-tasks) and that feels gross in comparison.

**Comment by bitprophet** [comment]
Grump, I was all set on `@task(triggers=[call, me, after])` but realized it's also ambiguous - I meant it as "Calling me _triggers_ calls to these _other_ tasks", but it could be read as a plural of "trigger", implying "other tasks which, when called, trigger a call _to me_".

Maybe `afterwards` is best for now after all? I should do some Twitter polls or something :)

**Comment by bitprophet** [comment]
More name ideas stemming from a (side) twitter discussion:

- Consider not cramming the same word into kwarg and term both, but using different ones
- Maybe I should stick with `@task(pre=[xxx], post=[yyy])` but have the terminology change from *pre-tasks* and *post-tasks* to *prerequisites* and *postrequisites*. Matches `make` to a degree which is both good and bad. Also just means more symmetry.
  - Counter: on their own, pre/post do suffer from the same ambiguity as many other kwargs, am I saying "xxx comes before this task" _or_ "this task comes before ('is pre-') xxx"?
- The current "dependencies / followups" terminology I'm using in the doc still seems like it might strike best balance between (lack of) ambiguity, clarity, ability to describe them in English, etc. Will have to see.

**Comment by offbyone** [comment]
`dependencies`/`consumers` is more explicit.

`@task(and_then=CONSUMERS)` works, as a documentation slug and name thingy.

(As an aside, the tendency of kwargs' names and storage to share a name in languages that support them really makes this kind of API a pain. Objective-C does this a lot better.)

**Comment by bitprophet** [comment]
`consumers` isn't necessarily generic enough, though; the example use cases that come up a lot are things like cleaning artifacts, notifying external services, etc. They aren't necessarily _consuming_ something _produced_ by the 'main' task (as opposed to e.g. a link-compile or static-asset pipeline where "consumer" is definitely appropriate.)

(One could make the argument that e.g. a notification followup task is "consuming" the [empty] output of e.g. a test or build task, but that feels like a stretch to me.)

`dependencies` seems like an obvious slam dunk either way though. I'm even contemplating adding an alias or two for it, e.g. `@task(requires=[other, tasks])`, but so far I've tried to shy away from having too many "convenient" aliases.

**Comment by bitprophet** [comment]
What if I want `check_versions` to run before some number of other tasks, instead of just `@compile`, though? :grin: Then we end up with this:

``` python
@task
def compile(): ...

@task
def build(): ...

@task
def dryrun(): ...

# ...

@compile.before
@build.before
@dryrun.before
# ...
def check_versions(): ...
```

**Comment by ask** [comment]
Yeah, I guess out of the question if these are supposed to build a tree of tasks. You can call other tasks in these, but it will be impossible to introspect what the dependencies are.   The good news I may take use of this pattern some time :)

**Comment by bitprophet** [comment]
Also just not sure how I feel about task objects being usable as decorators; it definitely makes for a neat-looking API, but I worry it goes too far into the magic zone with not enough benefit. Just a gut feeling though. (EDIT: but yea, I can definitely see other use cases where the benefits _do_ outweigh the drawbacks, so, good luck :D)

**Comment by ask** [comment]
@bitprophet The task objects are not decorators, that would the composite `@task.after` etc. 

```
class Task:

   def __init__(self, ...) -> None:
     self.before = Callbacks()
     self.after = Callbacks()

class Callbacks(MutableSequence, Callable):

    def __call__(self, fun: Callable) -> Callable:
        self.append(fun)
        return fun
```
I wouldn't call it magic exactly, implementation is simple, and It's used in the stdlib with `@property`:

```py
class X:
    _foo = None
    @property
    def foo(self):
        return self._foo

    @foo.setter
    def foo(self, value):
        self._foo = value

    @foo.deleter
    def foo(self):
         print('OOPS')
```

(Alas, I cannot argue that `@property` is good use of it.)

**Comment by bitprophet** [comment]
Ah right, good point. (Sorry, bouncing all over the place right now so even more scatterbrained than usual.)

My other point still stands unfortunately, I think it makes more sense in the general case for the declarations to live in the tasks making them instead of vice versa. This overall problem space, of course, often sees solutions going both ways (see eg Chef resources' `notifies`/`subscribes`) but I think I'd prefer to implement the more commonly useful variant first, and leave the option on the table to add the inverse later if enough people seem to want it.

**Comment by bitprophet** [comment]
Note to self: while writing out multiple examples using `@task(followups=[a,b,c])`,  suspect it should really be `@task(followup=[a,b,c])` instead.

----

Also noting FTR that Twitter has yielded more "literate" style ideas for the kwargs, e.g.:

- `@task(then=[more, stuff])`
- `@task(and_then=[do, more, things])`
- `@task(but_first=[do, this], afterwards=[even, more])`

My gut says these are slightly too cutesy, but you never know.

**Comment by offbyone** [comment]
Don't undersell it; there's definitely room for a sense of play in an API if the API is still usable. Especially if the literate approach yields clarity.

**Comment by bitprophet** [comment]
First pass at DDD is done, Github rendering of it is here: https://github.com/pyinvoke/invoke/blob/34c71cad54508579698bc5200dfaf3e65dd32eb5/sites/docs/concepts/execution.rst

Still says "followups" for now, I'll figure out what the final terminology should be after I actually prove an implementation works...

**Comment by bitprophet** [comment]
Each time I touch this stuff I find the API design parts irritating. Currently torn on whether to aim for multiple kwargs for the singular/plural case, or a single kwarg that behaves in a polymorphic fashion (accepts one object or an iterable of them).

E.g. `@task(dependencies=[a,b,c])` + `@task(depends_on=a)`, or...just `@task(depends_on=[a,b,c])` + `@task(depends_on=a)`. (I.e. in English, one can "depend on" a singular or plural noun, so...why not just go with that? A lot more options here work in the either-or case than are purely singular or purely plural.)

Also still torn between `depends_on` and `requires`. (Either could work as a polymorphic kwarg.)

**Comment by thebjorn** [comment]
`depends_on` and `requires` read much better than `dependencies` (which is too long and has too many syllables).

Having an easy type signature will let IDEs help programmers spot errors, and it would prevent gymnastics if one of the required pre-requisites is an iterable, so I would urge always using a list, i.e. `@task(requires=[a])`.

`afterwards` doesn't seem too cutesy (or maybe `on_success`/`on_error`..? -- I don't recall if post_tasks run regardless of task success/failure).

**Comment by bitprophet** [comment]
Agree that shorter is better.

While I recognize your assertions about signature/gymnastics, I'm eternally on the fence about "always a list" because it's an annoying chore in the _very_ common case of only ever wanting to throw a single value into it, and what's this sort of code for if not removing annoying chores? :grin: But again, I recognize the issues with polymorphism (or w/e it's properly called) which is why I wonder if there are any useful terms that strongly imply only the singular. Can't think of any really.

Then again - because we already do some mild gymnastics for the positional-argument use case, that makes 'single dependency' trivially easy (`@task(my_dependency)`) so perhaps it's just moot.

Re: success vs error, my guiding principle right now is "if a given logic doesn't _require_ the dependency system to achieve, it should not be implemented in that system at all", and I can't think of many scenarios where an `on_error` makes _more_ sense than some in-task `try`/`except`/`else`/`finally` construct.

"Normal" dependencies & followups make sense because it's not possible (certainly not easy) to achieve dependency deduplication or followup deferment without a call graph system; but anything where your logic dictates you want something to _always_ happen, and _immediately_ before/after the main task, seems like it should by rights live within the task body.

**Comment by bitprophet** [comment]
Hrm. How about `enables` for followup tasks (`@task(requires=[clean], enables=[notify])`)? Not great, since "you _can_ do X after you run me" is very different from "you _must_ do X after you run me". But it's another one for the brainstorm pile, and it seems to match up well with `requires` at least.

**Comment by offbyone** [comment]
`enables` is an ordering constraint, but doesn't state that they *will* be executed after. It'd be useful for a statement that those tasks *must* follow this one, but not that they *will*.

**Comment by bitprophet** [comment]
Isn't that what I said? :D

**Comment by bitprophet** [comment]
Moar: while they feel too mouthy/awkward, specifying dependencies with `preceded_by` and 
followups with `succeeded_by` at least has the property of being symmetrical:

``` python
@task(preceded_by=[clean], succeeded_by=[notify])
```

Yea...definitely awkward to type.

Alternately, flip it around and specify dependencies with `succeeds` and followups with `precedes`? Bit easier to type (if still having the double-c and double-e. boy we're in the weeds now aren't we?)

``` python
@task(succeeds=[clean], precedes=[notify])
```

**Comment by bitprophet** [comment]
Spent the time to gather up _all_ of the stupid words we've all brainstormed, and my personal thoughts/observations on 'em, and put them in the description. Please ping me if I missed your favorite 😛

**Comment by bitprophet** [comment]
The more I stare at and/or enhance these lists the more I feel like a) the main limiter is the "tasks after this task" plural noun, everything else has half decent options, and b) the least-awful of those is still "followups".

So I guess I'll keep rolling with that for now, and also keep referring to the overall system as "dependencies" or "the dependency system" since I still suspect dependencies will always be the focus by far.

**Comment by bitprophet** [comment]
Taking a stab at resurrecting this in a post-1.0 world. Sadly it means some of the "clean" changes now have to content with backwards compatibility, though I think that only really means keeping some arg/flag aliases around.

My old branch is nowhere near a clean merge due to a lot of the cleanup, file renaming & file consolidation that happened in the last year and a quarter (uggh) so I'm gonna have to do a lot of copy-pasting into a new branch or something. What needs doing:

- [ ] Port over the latest copy of the old execution doc into the modern invoking-tasks doc, so the DDD is preserved. Basically just load the latest versions of both up side by side and see what needs copying.
- [ ] Ditto the tests, which ought to be a little easier since it's _mostly_ file renaming and then additions only.
  - any modifications to existing tests should get preserved as having both, for backwards compat reasons
- [ ] Make sure backwards compatibility re: pre/post args is preserved (should still map cleanly enough to the new concepts)
- [ ] Ditto the `--no-dedupe` CLI flag, requires figuring out whether it should disable the newer system entirely, or if it should only impact the same aspects as it did before.
- [ ] Reread the new diff to make sure it's consistent & I'm still happy with the direction...
  - if I'm unlucky it will feel like it "really wants" to stay backwards incompat, which would be problematic & require a rethink of how to cleanly implement it
  - e.g. perhaps a new opt-in executor and/or program subclass users can enable to get the new jazz

**Comment by bitprophet** [comment]
In re-reading the docs I'm finding the hard requirement of an iterable value for `depends_on`/`afterwards` a little annoying/weird, especially since the singular `check` exists. Don't see that it'll make life _that_ much harder to allow the former two to be iterable-or-callable; esp since they don't (currently) accept strings (though we probably want to pre-emptively guard against that anyways).

Kinda wish `check`/`checks` had a nicer single name we could do the same for, since then it feels different from its siblings. Might have to double check my old brainstorm.

**Comment by pombredanne** [comment]
ping :) @bitprophet is this dead dead or... worthy of a resurrection?

**Comment by mohnen** [comment]
I know that I am late to the show, but have a look at [pyinvokedepends](https://pypi.org/project/pyinvokedepends/)
